src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs

```diff
@@ -520,6 +545,25 @@ protected override AuthenticationTicket AuthenticateCore()
             }
         }
 
+        protected virtual async Task<TokenResponse> RedeemAuthorizationCode(string authorizationCode)
+        {
+            var tokenRequestParameters = new Dictionary<string, string>()
+            {
+                { "client_id", Options.ClientId },
+                { "redirect_uri", Options.RedirectUri },
+                { "client_secret", Options.ClientSecret },
+                { "code", authorizationCode },
+                { "grant_type", "authorization_code" },
+            };
+            var requestMessage = new HttpRequestMessage(HttpMethod.Post, _configuration.TokenEndpoint);
+            requestMessage.Content = new FormUrlEncodedContent(tokenRequestParameters);
```

Comment:
As a side note, that would be great to have new notifications to alter the token request before it is sent and after the response is received. 

I already planned to submit a PR to add these hooks to the OAuth2 handler when we are done with the other PR.